src/utils/recompiler.ts

```diff
@@ -0,0 +1,201 @@
+            outputPath: newCacheEntry[cairoContract] ? newCacheEntry[cairoContract].outputPath : outputPath,
+            abiPath: newCacheEntry[cairoContract] ? newCacheEntry[cairoContract].abiPath : abiPath
+        };
+    }
+
+    return newCacheEntry;
+};
+
+// Gets cache entry of a given cairo file plus artifacts
+const getHashEntry = async (
+    file: string,
+    output: string,
+    abi: string
+): Promise<CacheEntry> => {
+    const data = await fsPromises.readFile(file);
+    const hash = createHash("sha256");
+    hash.update(data);
+
+    const newCacheEntry: CacheEntry = {};
+    newCacheEntry[file] = {
+        contentHash: hash.digest("hex").toString(),
+        outputPath: output,
+        abiPath: abi
+    };
+
+    return newCacheEntry;
+};
+
+// Updates cache entry with new contracts
+const getUpdatedCashEntry = (
+    oldCacheEntry: CacheEntry,
+    newCacheEntry: CacheEntry
+): CacheEntry => {
+    const updatedCacheEntry: CacheEntry = oldCacheEntry;
+    for (const contractName in newCacheEntry) {
+        if (oldCacheEntry[contractName]?.contentHash !== newCacheEntry[contractName].contentHash) {
+            updatedCacheEntry[contractName] = newCacheEntry[contractName];
+        }
+    }
+
+    return updatedCacheEntry;
+};
+
+// Checks artifacts availability
+const checkArtifacts = async (
+    hre: HardhatRuntimeEnvironment,
+    newCacheEntry: CacheEntry
+): Promise<Set<string>> => {
+    // Set to save contracts with changed content & unavailable artifacts
+    const changed: Set<string> = new Set();
+    const { starknetSources: defaultSourcesPath } = hre.config.paths;
+    const artifactsDir = getArtifactPath(defaultSourcesPath, hre);
+    // traverse on artifacts directory
+    const artifactsList = await traverseFiles(artifactsDir, "*.json");
+    for (const name in newCacheEntry) {
+        const outputPath = newCacheEntry[name].outputPath;
+        const abiPath = newCacheEntry[name].abiPath;
+        if (!artifactsList.includes(outputPath) || !artifactsList.includes(abiPath)) {
+            changed.add(name);
+        }
+    }
+
+    return changed;
+};
+
+// Compile changed contracts
+const compileChangedContracts = async (
+    args: TaskArguments,
+    hre: HardhatRuntimeEnvironment,
+    newCacheEntry: CacheEntry,
+    oldCacheEntry: CacheEntry,
+    changed: Set<string>
+): Promise<void> => {
+    for (const contractName in newCacheEntry) {
+        // Add new contracts that are not in cache before
+        if (!oldCacheEntry[contractName]) {
+            changed.add(contractName);
+        }
+
+        // Add contracts that contiain a change in content
+        if (newCacheEntry[contractName].contentHash !== oldCacheEntry[contractName]?.contentHash) {
+            changed.add(contractName);
+        }
+    }
+
+    if (changed.size > 0) {
+        // Compiles contracts
+        console.log("Compiling contracts...");
+        const copyPaths = args.paths;
```

Comment:
Why should the the `args` be passed from the other task? Like, why should args intended for e.g. `starknet-invoke` be given to `starknet-compile`? A new `args` object should be constructed here. Not sure if `compileChangedContracts` even needs an `args` parameter.

Reply:
https://github.com/Shard-Labs/starknet-hardhat-plugin/blob/3c16ef595ce68cf479a1a72f3bee82e182226a25/src/recompiler.ts#L165